### PR TITLE
search-input task

### PR DESCRIPTION
**Pull request by kovpaka**

### CODE REVIEW

File: src/components/SearchInputTask/SearchInputTask.jsx [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import SearchInput from '../SearchInput/SearchInput.jsx';
+
+const SearchInputModes = {
```
Comment (siarheikarabko) [inline src/components/SearchInputTask/SearchInputTask.jsx:4]: Please, move it to the component `SearchInput.jsx` and export this mode
`export SearchInputModes = { ... }`
and then import it here along with the component.
`import SearchInput, { SearchInputModes } from ../SearchInput/SearchInput.jsx';`

File: src/components/SearchInputTask/SearchInputTask.jsx [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    AfterStopedTyping: 'AfterStopedTyping',
+};
+
+export default function SearchInputTask() {
+    function onSearchHandler(searchText) {
+        console.log(searchText);
+    }
+
+    return (
+        <div className="search-input-task">
```
Comment (siarheikarabko) [inline src/components/SearchInputTask/SearchInputTask.jsx:16]: Do you need this classname?

File: src/components/SearchInput/SearchInput.jsx [hunk 3]
```diff
@@ -0,0 +1,66 @@
+import React from 'react';
+
+export default function SearchInput(props) {
+    let timer;
+    const doneTypingInterval = 2000;
```
[the remark this answers is not in the document]
Comment (siarheikarabko) [reply src/components/SearchInput/SearchInput.jsx:5]: Please, read these articles about debouncing and throttling.
https://redd.one/blog/debounce-vs-throttle
https://ellenaua.medium.com/throttle-debounce-behavior-lodash-6bcae1494e03
You could try to use `throttle` in `"Immediate"` mode and  `debounce` in `"AfterStopedTyping"` mode (use `lodash` lib or create your own implementation).

File: src/components/SearchInputTask/SearchInputTask.jsx (hunk 1 above)
Comment (kovpaka) [reply src/components/SearchInputTask/SearchInputTask.jsx:4]: Fixed

File: src/components/SearchInputTask/SearchInputTask.jsx (hunk 2 above)
Comment (kovpaka) [reply src/components/SearchInputTask/SearchInputTask.jsx:16]: Ok, removed

File: src/components/SearchInput/SearchInput.jsx [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                    onInput={handleInput}
+                />
+            );
+            break;
+        case 'AfterEnter':
+            searchInput = (
+                <input
+                    type="text"
+                    placeholder={props.placeholder}
+                    onKeyPress={handleKeyPress}
```
[the remark this answers is not in the document]
Comment (kovpaka) [reply src/components/SearchInput/SearchInput.jsx:47]: Fixed

File: src/components/SearchInput/SearchInput.jsx [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        props.onSearch(e.target.value);
+    }
+
+    function handleKeyPress(e) {
+        if (e.key === 'Enter') {
+            props.onSearch(e.target.value);
+        }
+    }
+
+    function handleChange(e) {
```
[the remark this answers is not in the document]
Comment (kovpaka) [reply src/components/SearchInput/SearchInput.jsx:17]: My bad, removed.

File: src/components/SearchInput/SearchInput.jsx [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+                    onKeyPress={handleKeyPress}
+                />
+            );
+            break;
+        case 'AfterStopedTyping':
+            searchInput = (
+                <input
+                    type="text"
+                    placeholder={props.placeholder}
+                    onKeyUp={handleKeyUp}
```
[the remark this answers is not in the document]
Comment (kovpaka) [reply src/components/SearchInput/SearchInput.jsx:56]: Ok

File: src/components/SearchInput/SearchInput.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (kovpaka) [reply src/components/SearchInput/SearchInput.jsx:5]: Done

File: src/components/SearchInput/SearchInput.jsx [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    function handleKeyDown() {
+        clearTimeout(timer);
+    }
+
+    let searchInput;
+
+    switch (props.mode) {
+        case 'Immediate':
+            searchInput = (
+                <input
```
[the remark this answers is not in the document]
Comment (kovpaka) [reply src/components/SearchInput/SearchInput.jsx:35]: Fixed

File: src/components/SearchInput/SearchInput.jsx [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+                    type="text"
+                    placeholder={props.placeholder}
+                    onKeyUp={handleKeyUp}
+                    onKeyDown={handleKeyDown}
+                />
+            );
+            break;
+        default:
+            throw new Error(`Unknown ${props.mode} mode`);
```
[the remark this answers is not in the document]
Comment (kovpaka) [reply src/components/SearchInput/SearchInput.jsx:62]: Search after enter is pressed is a default search now, if you don't mind.

**Pull request merged**